### ISSUE TITLE
Huge drops in write throughput correlated with compaction

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#1911]
*Installation details*
Scylla version (or git commit hash): master
Cluster size: 1 node, -c1 -m1G

```
cassandra-stress write no-warmup duration=60m -pop seq=1..1000000
```

Drop is from 25K to 0.9K. Drop happens periodically. There are short drops and long drops.

Drops correlate with CPU user time reaching 100% and active compaction.

[io-dash.pdf](https://github.com/scylladb/scylla/files/624436/io-dash.pdf)
[per-server-dash.pdf](https://github.com/scylladb/scylla/files/624435/per-server-dash.pdf)

**Comment by tgrabiec** [comment]
Cannot tell what's consuming CPU exactly due to #1909

**Comment by tgrabiec** [comment]
Disabling compaction makes the big drops go away, leaving us with small drops (25K -> 17K) correlated with memtable flush.

**Comment by tgrabiec** [comment]
Running compaction and memtables under scheduling groups makes the workload much smoother.

Before:

![screenshot from 2016-12-01 18-23-49](https://cloud.githubusercontent.com/assets/283695/20804546/a003891c-b7f3-11e6-966d-99241d200165.png)


After:

![screenshot from 2016-12-01 18-23-17](https://cloud.githubusercontent.com/assets/283695/20804552/a6b2b3fa-b7f3-11e6-8382-a7f016237b88.png)


compaction: period=10ms, quota=20%
flush: period=2ms, quota=10%

**Comment by raphaelsc** [comment]
BTW, I think it would be important to use different scheduling groups for sstable creation as a result of compaction and memtable flush. That's for workload conditioning where we may need to give more CPU for compaction when it falls behind. Compaction of sstables with lots of small partitions is likely to be bound by CPU. @glommer

**Comment by tgrabiec** [comment]
The concurrency used by the client is enough to saturate the CPU. Latency of writes is low when throughput is high (order of microseconds), so processing of the whole "batch" of writes from all concurrent clients takes little time. When flush/compaction tasks start, they can consume the CPU continuously for 2ms on average until they are preempted. This will delay polling and processing of incoming writes. Because the batch takes short time to process, only a fraction of time will be dedicated by the server to processing of writes.

Putting the CPU hogs under a scheduling group improved the situation because their CPU consumption within given period was limited to 10-20%, allowing writes to be served without delay for the rest of the period. 

This is not a solution to the problem though because there are times when we would want flush and compaction to be given more CPU, otherwise priority inversion problem would happen. E.g. writes (high priority) would block on dirty memory, and thus indirectly on the flush.

**Comment by avikivity** [comment]
Fixed by cce1a2bce868d19034c2de50c1ace4546e0244ad?

**Comment by avikivity** [comment]
Assumed fixed by cce1a2bce868d19034c2de50c1ace4546e0244ad, please reopen if this is not the case.
